warehouse/query-core/src/main/java/org/apache/commons/jexl2/parser/ASTDelayedPredicate.java

```diff
@@ -59,4 +67,48 @@ public static ASTDelayedPredicate create(JexlNode node) {
         
         return expr;
     }
+    
+    /**
+     * Recursively ascend the tree looking for an instance of an ASTDelayedPredicate intersected with a parent
+     * 
+     * @param node
+     *            the node to check
+     * @return true if this node is delayed, or if a parent node is delayed
+     */
+    public static boolean isSubTreeAlreadyDelayed(JexlNode node) {
+        if (node instanceof ASTAndNode && node.jjtGetNumChildren() == 2) {
+            if (QueryPropertyMarker.findInstance(node).isType(ASTDelayedPredicate.class)) {
+                return true;
+            }
+        }
+        
+        if (node.jjtGetParent() == null) {
+            return false;
+        } else {
+            return isSubTreeAlreadyDelayed(node.jjtGetParent());
+        }
+    }
+    
+    /**
+     * Unwrap a delayed predicate, fully. Intended to handle the odd edge case when multiple delayed predicate markers are applied to the same node
+     *
+     * @param node
+     *            an arbitrary jexl node
+     * @return the source node, or the original node if this node is not delayed
+     */
+    public static JexlNode unwrapFully(JexlNode node) {
+        if (QueryPropertyMarker.findInstance(node).isType(ASTDelayedPredicate.class)) {
```

Comment:
Not sure how often we'll be doing this or how deep these trees can grow but you could save a traversal by doing:

```
Instance instance = QueryPropertyMarker.findInstance(node);
if (instance.isType(ASTDelayedPredicate.class)) {
  JexlNode source = instance.getSource();
  // ...
```